File: forge/features/middleware/templates/security_headers/python/files/src/app/middleware/security_headers.py
"""Security-headers middleware.

Adds a conservative set of response headers that mitigate common browser-side
attack classes. Defaults are safe for APIs served from a Traefik-fronted stack;
tune CSP/HSTS when this service also serves HTML.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable

from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware

_DEFAULT_HEADERS: dict[str, str] = {
    "X-Content-Type-Options": "nosniff",
    "X-Frame-Options": "DENY",
    "Referrer-Policy": "strict-origin-when-cross-origin",
    "Permissions-Policy": "accelerometer=(), camera=(), geolocation=(), microphone=()",
    # CSP restricted to same-origin by default. Relax for HTML/asset-serving services.
    "Content-Security-Policy": "default-src 'self'; frame-ancestors 'none'",
    # HSTS: send only when the request arrived over TLS. Browsers ignore HSTS
    # over plain HTTP anyway, but we avoid noise in the header set.
}


class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """Inject standard security headers into every response."""

    def __init__(
        self,
        app,
        *,
        extra_headers: dict[str, str] | None = None,
        hsts_max_age: int = 31536000,
    ):
        super().__init__(app)
        self._headers = dict(_DEFAULT_HEADERS)
        if extra_headers:
            self._headers.update(extra_headers)
        self._hsts_max_age = hsts_max_age

    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        response = await call_next(request)
        for name, value in self._headers.items():
            response.headers.setdefault(name, value)
        if request.url.scheme == "https":
            response.headers.setdefault(
                "Strict-Transport-Security",
                f"max-age={self._hsts_max_age}; includeSubDomains",
            )
        return response

File: forge/features/middleware/templates/security_headers/rust/files/src/middleware/security_headers.rs
//! Security-headers middleware.
//!
//! Injects a conservative set of response headers on every response. Matches
//! the Python fragment's defaults: nosniff, frame DENY, strict referrer,
//! restrictive CSP, restrictive Permissions-Policy, and HSTS when served over
//! TLS (HSTS is omitted from plain-HTTP responses — browsers ignore it anyway).

use axum::body::Body;
use axum::http::{HeaderName, HeaderValue, Request};
use axum::middleware::Next;
use axum::response::Response;

const HEADERS: &[(&str, &str)] = &[
    ("x-content-type-options", "nosniff"),
    ("x-frame-options", "DENY"),
    ("referrer-policy", "strict-origin-when-cross-origin"),
    (
        "permissions-policy",
        "accelerometer=(), camera=(), geolocation=(), microphone=()",
    ),
    (
        "content-security-policy",
        "default-src 'self'; frame-ancestors 'none'",
    ),
];

pub async fn security_headers_middleware(req: Request<Body>, next: Next) -> Response {
    let is_https = req.uri().scheme_str() == Some("https");
    let mut response = next.run(req).await;
    let headers = response.headers_mut();
    for (name, value) in HEADERS {
        if let (Ok(h), Ok(v)) = (
            HeaderName::from_bytes(name.as_bytes()),
            HeaderValue::from_str(value),
        ) {
            headers.entry(h).or_insert(v);
        }
    }
    if is_https {
        headers
            .entry(HeaderName::from_static("strict-transport-security"))
            .or_insert(HeaderValue::from_static(
                "max-age=31536000; includeSubDomains",
            ));
    }
    response
}
